_posts/2015-02-20-elasticsearch-and-denormalization.md

```diff
@@ -0,0 +1,262 @@
+view. The `SampleDenormalizer` looks like this:
+
+[presenter-pattern]: http://nithinbekal.com/posts/rails-presenters/ "Nithin Bekal on Presenters in Rails"
+
+```ruby
+class SampleDenormalizer
+
+  def initializer(sample)
+    @sample = sample
+  end
+
+  # This method should be treated like a view - no logic, just attributes
+  def to_hash
+    # this list of methods can get very long but it encourages you to
+    # keep the hash flat and to be deliberate about what you are adding to it
+    %w(id
+       display_name
+       original_image
+       style_id
+       style_name
+       color_name).map{ |method_name|
+        [ method_name, self.send(method_name) ]
+    }.to_h
+  end
+
+  def display_name
+    @sample.display_name
+  end
+
+  def original_image
+    @sample.original_image
+  end
+
+  def color_name
+    @sample.color.name
+  end
+
+  def style_name
+    style.name
+  end
+
+  def style_id
+    style.id
+  end
+
+  private
+  # any methods that aren't included in the hash can be private and should be
+  # memoized for better performance
+  def style
+    @style  ||= @sample.style
+  end
+
+end
+```
+
+You can now use the denormalizer in the model to build the hash for
+ElasticSearch, considerably reducing the complexity of this particular class.
+
+```ruby
+class Sample < ActiveRecord::Base
+  include Elasticsearch::Model
+
+  def to_hash
+    SampleDenormalizer.new(self).to_hash
+  end
+  
+  # mapping here as before ...
+
+end
+```
+
+One of the strengths of this pattern is for calculated attributes like
+`has_fit_attributes?` which returns a boolean value but is complex and expensive
+to calculate. What's more, it is nearly impossible to query with SQL. Once you
+put it in the ElasticSearch index it is cheap to store and return and simple to
+query.
+
+```ruby
+class SampleDenormalizer
+
+  def to_hash
+    # added `has_fit_attributes?` to the hash
+    %w(id
+       display_name
+       original_image
+       style_id
+       style_name
+       has_fit_attributes? 
+       color_name).map{ |method_name|
+        [ method_name, self.send(method_name) ]
+    }.to_h
+  end
+
+  # same as example above ...
+
+  def has_fit_attributes?
+    return true if @sample.root_item_type  == 'Non-Apparel'
+    required_scalar_values  = %w{fit_top good_for_top_body_type}.map{|attr| style.scalar_attributes.send(attr) }
+    required_multi_values   = %w{bra_type shoulder_width_match torso_length_match}.map do |attr|
+      style.multi_attributes_names.include?(attr)
+    end
+    (required_scalar_values + required_multi_values).all?
+  end
+
+end
+```
+
+So that's how you can use the Denormalizer pattern to build up your search
+index. You can use that same code for more typical presenter work.
+
+```ruby
+class SamplesController < ApplicationController
+
+  # the objects being presented here are Hashie::Mash objects from ElasticSearch
+  def index
+    @samples  = Sample.search(params[:keywords]).map{|s| SamplePresenter.new(s._source) }
```

Comment:
This confused me a bit.  Where is `search` defined and what is going on with the `_source` method?

You could break up this section into two—in the first, you say "our `index` method uses the ES-provided `search` method.  We wrap the `_source` (which is a blah blah blah) in our `SamplePresenter` (which we'll see in a minute)".  In the second ,you say "our `update` method, which operates on the `Sample` AR object **also** gets wrapped in our `SamplePresenter` <<show code>> What's up with that?!"

This takes smaller steps and also builds to the punch line a bit more explicitly